File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/account/AddonUrlNormalizer.kt
package com.crispy.tv.domain.account

sealed interface AddonUrlResult {
    data class Valid(val normalized: String) : AddonUrlResult
    data object Invalid : AddonUrlResult
}

fun normalizeAddonUrl(raw: String): AddonUrlResult {
    val trimmed = raw.trim()
    if (!trimmed.startsWith("https://", ignoreCase = true)) return AddonUrlResult.Invalid
    val lower = trimmed.lowercase()
    val withoutFragment = lower.substringBefore('#')
    val noTrailingSlash = withoutFragment.removeSuffix("/")
    val path = noTrailingSlash.substringAfter("https://").substringBefore('?')
    if (!path.contains('/')) return AddonUrlResult.Invalid
    val segments = path.split('/').filter { it.isNotEmpty() }
    if (segments.isEmpty()) return AddonUrlResult.Invalid
    val last = segments.last()
    if (!last.endsWith(".json")) return AddonUrlResult.Invalid
    return AddonUrlResult.Valid(noTrailingSlash)
}

File: ios/ContractRunner/Sources/ContractRunner/AddonUrlNormalizer.swift
import Foundation

public enum AddonUrlResult {
    case valid(String)
    case invalid
}

public func normalizeAddonUrl(_ raw: String) -> AddonUrlResult {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard trimmed.lowercased().hasPrefix("https://") else {
        return .invalid
    }
    let lower = trimmed.lowercased()
    let withoutFragment = lower.split(separator: "#").first.map(String.init) ?? lower
    let noTrailingSlash = withoutFragment.hasSuffix("/")
        ? String(withoutFragment.dropLast())
        : withoutFragment
    let afterHost = noTrailingSlash.dropFirst("https://".count)
    let path = afterHost.split(separator: "?").first.map(String.init) ?? String(afterHost)
    let segments = path.split(separator: "/").filter { !$0.isEmpty }
    guard let last = segments.last, last.hasSuffix(".json") else {
        return .invalid
    }
    return .valid(noTrailingSlash)
}
